browser/ephemeral_storage/ephemeral_storage_qa_browsertest.cc

```diff
@@ -0,0 +1,977 @@
+                                   AsString(expected[2]) + "') !== null"));
+    ASSERT_EQ(true,
+              EvalJs(contents, "document.querySelector('#storage-rs .row-" +
+                                   AsString(storage_type) +
+                                   " .cell-nested-frame.bg-" +
+                                   AsString(expected[3]) + "') !== null"));
+  }
+
+  // Corresponds to step 1 of QA test page
+  void TestInitialCase(const StorageResult expected[4][4]) {
+    ASSERT_TRUE(original_tab_);
+
+    CheckStorageResults(original_tab_, expected);
+  }
+
+  // Corresponds to step 2 of QA test page
+  void TestRemotePageSameSession(const StorageResult expected[4][4]) {
+    ASSERT_TRUE(original_tab_);
+    ASSERT_EQ(1, tabs_->count());
+
+    NavigateOtherOrigin(original_tab_);
+    ASSERT_EQ(2, tabs_->count());
+    ASSERT_EQ(1, tabs_->active_index());
+
+    content::WebContents* contents = tabs_->GetActiveWebContents();
+
+    ClickReadValues(contents);
+
+    CheckStorageResults(contents, expected);
+  }
+
+  // Corresponds to step 3 of QA test page
+  void TestRemotePageNewSession(const StorageResult expected[4][4]) {
+    ASSERT_TRUE(original_tab_);
+    ASSERT_EQ(1, tabs_->count());
+
+    chrome::NewTab(browser());
+    ASSERT_EQ(2, tabs_->count());
+    ASSERT_EQ(1, tabs_->active_index());
+
+    std::string target =
+        EvalJs(original_tab_,
+               "document.getElementById('continue-test-url-step-3').value")
+            .ExtractString();
+    ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), GURL(target)));
+
+    content::WebContents* contents = tabs_->GetActiveWebContents();
+
+    ClickReadValues(contents);
+
+    CheckStorageResults(contents, expected);
+  }
+
+  // Corresponds to step 4 of QA test page
+  void TestThisPageSameSession(const StorageResult expected[4][4]) {
+    ASSERT_TRUE(original_tab_);
+    ASSERT_EQ(1, tabs_->count());
+
+    NavigateSameOrigin(original_tab_);
+    ASSERT_EQ(2, tabs_->count());
+    ASSERT_EQ(1, tabs_->active_index());
+
+    content::WebContents* contents = tabs_->GetActiveWebContents();
+
+    ClickReadValues(contents);
+
+    CheckStorageResults(contents, expected);
+  }
+
+  // Corresponds to step 5 of QA test page
+  void TestThisPageDifferentSession(const StorageResult expected[4][4]) {
+    ASSERT_TRUE(original_tab_);
+    ASSERT_EQ(1, tabs_->count());
+
+    chrome::NewTab(browser());
+    ASSERT_EQ(2, tabs_->count());
+    ASSERT_EQ(1, tabs_->active_index());
+
+    std::string target =
+        EvalJs(original_tab_,
+               "document.getElementById('continue-test-url-step-5').value")
+            .ExtractString();
+    ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), GURL(target)));
+
+    content::WebContents* contents = tabs_->GetActiveWebContents();
+
+    ClickReadValues(contents);
+
+    CheckStorageResults(contents, expected);
+  }
+
+  // Corresponds to step 6 of QA test page
```

Comment:
please list the actual steps instead of referring to the page so other people don't have to toggle back and forth between .cc and .html to understand what is going on